### PR TITLE
sample data uses download instead of retry

### PR DESCRIPTION
**Pull request by nabobalis**
There might be a bug in parfive's retry function. We replace it for now with a new downloader.

### CODE REVIEW

**Comment by nabobalis** [comment]
```
KeyError: 'AIA_171_ROLL_IMAGE'
```
Not sure what to make of the RTD error.

**Comment by nabobalis** [comment]
It didn't error this time?

File: readthedocs.yml [hunk 1]
```diff
@@ -10,7 +16,6 @@ sphinx:
   fail_on_warning: true
 
 python:
-   version: 3.8
```
Comment (nabobalis) [inline readthedocs.yml:13]: Python version is governed by conda.

File: rtd-environment.yml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 name: sunpy
 channels:
-  - conda-forge
+  - defaults
```
Comment (nabobalis) [inline rtd-environment.yml:3]: We don't need conda-forge to install openjpeg.

File: sunpy/data/tests/test_sample.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import os
+
+import pytest
+
+from .._sample import download_sample_data
+
+
+@pytest.mark.remote_data
+def test_sample_download(tmpdir):
```
Comment (nabobalis) [inline sunpy/data/tests/test_sample.py:9]: Probably want to skip this by default.

**Comment by nabobalis** [comment]
```
DEBUG: Failed to download EVE_TIMESERIES from http://data.sunpy.org/sunpy/v1/20110607_EVE_L0CS_DIODES_1m.txt: Timeout on reading data from socket [sunpy.data._sample]
DEBUG: Will redownload using https://github.com/sunpy/sample-data/raw/master/sunpy/v1/20110607_EVE_L0CS_DIODES_1m.txt [sunpy.data._sample]
DEBUG: Failed to download EVE_TIMESERIES from https://github.com/sunpy/sample-data/raw/master/sunpy/v1/20110607_EVE_L0CS_DIODES_1m.txt: 400, message='Can not decode content-encoding: gzip' [sunpy.data._sample]
Task was destroyed but it is pending!
task: <Task pending name='Task-103' coro=<Downloader._write_worker() running at /home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/site-packages/parfive/downloader.py:469> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f24c65de5e0>()]>>
Exception ignored in: <coroutine object Downloader._write_worker at 0x7f24c65cc740>
Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/site-packages/parfive/downloader.py", line 469, in _write_worker
    offset, chunk = await queue.get()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/asyncio/queues.py", line 165, in get
    getter.cancel()  # Just in case getter is not done yet.
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/asyncio/base_events.py", line 719, in call_soon
    self._check_closed()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
```

File: sunpy/data/_sample.py [hunk 4]
```diff
@@ -78,50 +104,23 @@ def download_sample_data(overwrite=False):
...
-            return results
-
-    for err in results.errors:
-        file_name = err.filepath_partial().name
-        log.debug(f"Failed to download {_sample_files[file_name]} from {err.url}: {err.exception}")
-        log.error(
-            f"Failed to download {_sample_files[file_name]} from all mirrors, the file will not be available.")
-
-    return results
+    return sorted(results + already_downloaded)
```
Comment (nabobalis) [inline sunpy/data/_sample.py:126]: Not sure if this is correct at all.

File: sunpy/data/tests/test_sample.py (hunk 3 above)
Comment (Cadair) [reply sunpy/data/tests/test_sample.py:9]: yeah

File: sunpy/data/_sample.py [hunk 5]
```diff
@@ -22,35 +22,30 @@
 
 # the files should include necessary extensions
 _sample_data = {
-    # Do roll image first because it's the largest file.
```
Comment (Cadair) [inline sunpy/data/_sample.py:25]: Why the reordering? This comment is still relevant (it starts the download first and then the other 4 slots normally download the rest of the files while this one is running?)

File: sunpy/__init__.py [hunk 6]
```diff
@@ -22,7 +22,6 @@
 try:
     from .version import __version__
 except ImportError:
-    print("version.py not found, please reinstall sunpy.")
```
Comment (Cadair) [inline sunpy/__init__.py:25]: We should keep this or something like it.

File: readthedocs.yml (hunk 1 above)
Comment (Cadair) [reply readthedocs.yml:13]: Doesn't this make RTD setup a conda env with this version? If not should we put the python version in the rtd-env file so we are explicit about it?

Comment (nabobalis) [reply readthedocs.yml:13]: I can add it.

File: sunpy/data/_sample.py (hunk 5 above)
Comment (nabobalis) [reply sunpy/data/_sample.py:25]: I wanted it alphabetical

Comment (Cadair) [reply sunpy/data/_sample.py:25]: I mean *sure*, but this one should still be at the top.

Comment (nabobalis) [reply sunpy/data/_sample.py:25]: Undone.

File: sunpy/__init__.py (hunk 6 above)
Comment (nabobalis) [reply sunpy/__init__.py:25]: Undone.

File: sunpy/data/tests/test_sample.py (hunk 3 above)
Comment (nabobalis) [reply sunpy/data/tests/test_sample.py]: Removed.

File: sunpy/data/_sample.py [hunk 7]
```diff
@@ -8,8 +8,8 @@
 from sunpy.util.parfive_helpers import Downloader
 
 _base_urls = (
-    'http://data.sunpy.org/sunpy/v1/',
     'https://github.com/sunpy/sample-data/raw/master/sunpy/v1/',
+    'http://data.sunpy.org/sunpy/v1/',
```
Comment (nabobalis) [inline sunpy/data/_sample.py:12]: Do we want to swap this?

File: sunpy/data/_sample.py (hunk 4 above)
Comment (Cadair) [reply sunpy/data/_sample.py:127]: iirc, I don't think it matters what order this comes back in, but I would need to check

File: sunpy/data/_sample.py (hunk 7 above)
Comment (Cadair) [reply sunpy/data/_sample.py:12]: Questions, Questions. I think given we basically have unlimited github LFS hosting now, I am going to say yes, but we should probably re-review the situation before the 2.1 release.

File: sunpy/data/_sample.py (hunk 4 above)
Comment (nabobalis) [reply sunpy/data/_sample.py:127]: I meant more if I can a results to a already_downloaded.

**Comment by nabobalis** [comment]
RTD is still failing to download the sample data.

**Comment by nabobalis** [comment]
Docs have built this time but still got:
```
Running Sphinx v3.1.2
DEBUG: Failed to download EVE_TIMESERIES from https://github.com/sunpy/sample-data/raw/master/sunpy/v1/20110607_EVE_L0CS_DIODES_1m.txt: 400, message='Can not decode content-encoding: gzip' [sunpy.data._sample]
DEBUG: Will redownload using http://data.sunpy.org/sunpy/v1/20110607_EVE_L0CS_DIODES_1m.txt [sunpy.data._sample]
loading translations [en]... done
Task was destroyed but it is pending!
task: <Task pending name='Task-369' coro=<Downloader._write_worker() running at /home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/site-packages/parfive/downloader.py:469> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f4c72812e50>()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-375' coro=<Downloader._http_download_worker() running at /home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/site-packages/parfive/downloader.py:519> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f4c727adac0>()]> cb=[gather.<locals>._done_callback() at /home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/asyncio/tasks.py:758]>
Exception ignored in: <coroutine object Downloader._write_worker at 0x7f4c73ec3940>
Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/site-packages/parfive/downloader.py", line 469, in _write_worker
    offset, chunk = await queue.get()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/asyncio/queues.py", line 165, in get
    getter.cancel()  # Just in case getter is not done yet.
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/asyncio/base_events.py", line 719, in call_soon
    self._check_closed()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4379/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
```

File: sunpy/data/_sample.py [hunk 8]
```diff
@@ -40,24 +39,67 @@
...
+
+def _retry_sample_data(results):
+    # In case we have a broken file on disk, overwrite it.
+    dl = Downloader(overwrite=True, progress=False)
+    for err in results.errors:
+        file_name = err.filepath_partial().name
+        log.debug(
+            f"Failed to download {_SAMPLE_FILES[file_name]} from {err.url}: {err.exception}")
+        # Update the url to a mirror and requeue the file.
+        new_url = urljoin(_BASE_URLS[1], file_name)
```
Comment (Cadair) [inline sunpy/data/_sample.py:90]: Probably doesn't really matter, but this now limits us to two mirrors?

File: sunpy/data/_sample.py [hunk 9]
```diff
@@ -40,24 +39,67 @@
...
+        log.debug(f"Will redownload using {new_url}")
```
Comment (Cadair) [inline sunpy/data/_sample.py:91]:
```suggestion
        log.debug(f"Attempting redownload of {_SAMPLE_FILES[file_name]} using {new_url}")
```

**Review by Cadair — APPROVED** [review]

File: sunpy/data/_sample.py (hunk 8 above)
Comment (nabobalis) [reply sunpy/data/_sample.py:90]: It does but my head hurt trying to work out the logic for a loop. I can try to figure it out if you prefer.

Comment (Cadair) [reply sunpy/data/_sample.py:90]: nah, we can cross that bridge when we come to it.

**Comment by Cadair** [comment]
This is a lot nicer than the tripe I originally wrote :rofl:

**Comment by nabobalis** [comment]
I am not sure this really fixes anything tho.

**Comment by Cadair** [comment]
yeah maybe not, but it is still cleaner.

**Comment by nabobalis** [comment]
DRMS export error on RTD.

**Comment by nabobalis** [comment]
Worth backporting?

**Pull request merged**